### PR TITLE
ENH: Add support for min_count keyword for Resample and Groupby functions

### LINKED ISSUE #37768: BUG:  downsampling with last doesn't accept "min_count" keyword
I verified this on v 1.1.3. I am on miniconda, 1.1.4 is not available yet. There are ways to install probably but if I read documentation right, this feature should be around since 0.22

```
import numpy as np
import pandas as pd

index = pd.date_range(start='2020', freq='M', periods=6)
data = np.ones(6)
data[4:6] = np.nan
datetime = pd.Series(data, index)
period = datetime.to_period()
datetime
2020-01-31    1.0
2020-02-29    1.0
2020-03-31    1.0
2020-04-30    1.0
2020-05-31    NaN
2020-06-30    NaN
Freq: M, dtype: float64
```

```
datetime.resample('Q').sum(min_count=2)
2020-03-31    3.0
2020-06-30    NaN
Freq: Q-DEC, dtype: float64
```

```
datetime.resample('Q').last(min_count=3)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-bd5dfd934676> in <module>
----> 1 datetime.resample('Q').last(min_count=3)

~\miniconda3\lib\site-packages\pandas\core\resample.py in g(self, _method, *args, **kwargs)
    933 
    934     def g(self, _method=method, *args, **kwargs):
--> 935         nv.validate_resampler_func(_method, args, kwargs)
    936         return self._downsample(_method)
    937 

~\miniconda3\lib\site-packages\pandas\compat\numpy\function.py in validate_resampler_func(method, args, kwargs)
    395             )
    396         else:
--> 397             raise TypeError("too many arguments passed in")
    398 
    399 

TypeError: too many arguments passed in
```

#### Problem description
According to the documentation, .last() does accept the keyword "min_count", just like for example .sum()
where it works fine, see above

So I should not see the error above. The "min_count" is useful also for .last() if you have nans in your data and want to avoid the record that is not truly the last record in the segment.

Doc:
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.core.resample.Resampler.last.html#pandas.core.resample.Resampler.last)

#### Expected Output
```
2020-03-31    1.0
2020-06-30    NaN
```

#### Output of ``pd.show_versions()``
See above - verified with pd 1.1.3. I started the issue on my laptop with older pandas but finished with 1.1.3. Error message and pd.show_versions() is up to date

<details>

INSTALLED VERSIONS
------------------
commit           : db08276bc116c438d3fdee492026f8223584c477
python           : 3.7.6.final.0
python-bits      : 64
OS               : Windows
OS-release       : 8.1
Version          : 6.3.9600
machine          : AMD64
processor        : Intel64 Family 6 Model 62 Stepping 4, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.None

pandas           : 1.1.3
numpy            : 1.19.2
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.2.4
setuptools       : 50.3.1.post20201107
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : 3.2.1
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.19.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.1.3
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : None
numba            : 0.49.1

</details>

### LINKED ISSUE #37821: BUG: Groupby.last accepts min_count but implementation raises
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
df =  DataFrame({"a": [1, 2, 3, 4, 5, 6]})
df.groupby(level=0).last(min_count=1)

```

#### Problem description

This raises 
```
Traceback (most recent call last):
  File "/home/developer/.config/JetBrains/PyCharm2020.2/scratches/scratch_4.py", line 97, in <module>
    print(df.groupby(level=0).last(min_count=1))
  File "/home/developer/PycharmProjects/pandas/pandas/core/groupby/groupby.py", line 1710, in last
    return self._agg_general(
  File "/home/developer/PycharmProjects/pandas/pandas/core/groupby/groupby.py", line 1032, in _agg_general
    result = self._cython_agg_general(
  File "/home/developer/PycharmProjects/pandas/pandas/core/groupby/generic.py", line 1018, in _cython_agg_general
    agg_mgr = self._cython_agg_blocks(
  File "/home/developer/PycharmProjects/pandas/pandas/core/groupby/generic.py", line 1116, in _cython_agg_blocks
    new_mgr = data.apply(blk_func, ignore_failures=True)
  File "/home/developer/PycharmProjects/pandas/pandas/core/internals/managers.py", line 425, in apply
    applied = b.apply(f, **kwargs)
  File "/home/developer/PycharmProjects/pandas/pandas/core/internals/blocks.py", line 368, in apply
    result = func(self.values, **kwargs)
  File "/home/developer/PycharmProjects/pandas/pandas/core/groupby/generic.py", line 1067, in blk_func
    result, _ = self.grouper.aggregate(
  File "/home/developer/PycharmProjects/pandas/pandas/core/groupby/ops.py", line 594, in aggregate
    return self._cython_operation(
  File "/home/developer/PycharmProjects/pandas/pandas/core/groupby/ops.py", line 547, in _cython_operation
    result = self._aggregate(result, counts, values, codes, func, min_count)
  File "/home/developer/PycharmProjects/pandas/pandas/core/groupby/ops.py", line 608, in _aggregate
    agg_func(result, counts, values, comp_ids, min_count)
  File "pandas/_libs/groupby.pyx", line 906, in pandas._libs.groupby.group_last
AssertionError: 'min_count' only used in add and prod

Process finished with exit code 1
```

#### Expected Output

I would expect, that this function does not accept ``min_count`` as input (similar to ``Resample.last``) and raises the same error as described in #37768 
Additonally the docs need adjustments, because ``min_count`` is documented. This probably holds true for most of the functions mentioned in https://github.com/pandas-dev/pandas/issues/37768#issuecomment-727009192 too.

#### Output of ``pd.show_versions()``

<details>

master

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #37768
- [x] closes #37821
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This would add support for the documented keyword ```min_count``. If we decide to move forward with this, we have to decide what to do in the following case:
```
df = DataFrame({"a": [1] * 3, "b": [np.nan] * 3})
result = df.groupby("a").func(min_count=0)
```

``first`` and ``last`` return ``0.0`` which is fine I think and consistent with ``sum``, but ``min`` returns ``inf`` while ``max`` returns ``-inf``.

Default parameter is set to ``1`` to avoid backwards incompatible changes. We should keep this at least until 2.0

If we decide to not implement ``min_count`` here, we should adjust the docs and the function signatures for the groupby functions.

cc @rhshadrach

### CODE REVIEW

**Comment by rhshadrach** [comment]
Thanks for the PR @phofl. It seems to me that first, last, min, and max should all return nan even when `min_count` is 0. Sum behaves differently because the empty sum is 0 (likewise, the empty product is 1).

**Comment by rhshadrach** [comment]
> Default parameter is set to 1 to avoid backwards incompatible changes. We should keep this at least until 2.0

Do you perhaps mean -1? Otherwise, I don't follow.

**Comment by DiSchi123** [comment]
My 2 cents: min_count=0 is the same as not specifying a min_count. Min_count of x says you need to have at least x non-nan values in the bin to show a result, otherwise show nan. If you have a min_count of 0 isnt that like saying show me the sum, first etc regardless how many nans are there? Maybe I am missing something? 

But while it is important to decide what min_count=0 should show it is less important for the implementation, you would not specify that imo, because as mentioned it is like not specifying it. I am ok with any decision.

**Comment by phofl** [comment]
cc @rhshadrach  When we want to return nan in this case, we can keep -1 as default.

File: pandas/tests/groupby/aggregate/test_other.py [hunk 1]
```diff
@@ -638,3 +638,12 @@ def weird_func(x):
 
     result = df["decimals"].groupby(df["id1"]).agg(weird_func)
     tm.assert_series_equal(result, expected, check_names=False)
+
+
+@pytest.mark.parametrize("func", ["first", "last", "max", "min"])
+def test_min_count_implementation_min_max_first_last(func):
```
Comment (rhshadrach) [inline pandas/tests/groupby/aggregate/test_other.py:644]: aggregate is for testing the agg method on groupby, this belongs in missing. As an aside, I'm not seeing much in the way of `min_count` tests outside of categorical. How about name this as `test_min_count` and can extend the scope later (okay to do it here too).

File: pandas/_libs/groupby.pyx [hunk 2]
```diff
@@ -1057,7 +1056,7 @@ def group_nth(rank_t[:, :] out,
 
             for i in range(ncounts):
                 for j in range(K):
-                    if nobs[i, j] == 0:
+                    if nobs[i, j] < min_count or nobs[i, j] == 0:
```
Comment (rhshadrach) [inline pandas/_libs/groupby.pyx:1059]: Is this tested?

File: doc/source/whatsnew/v1.2.0.rst [hunk 3]
```diff
@@ -235,6 +235,7 @@ Other enhancements
 - Improve error reporting for :meth:`DataFrame.merge()` when invalid merge column definitions were given (:issue:`16228`)
 - Improve numerical stability for :meth:`Rolling.skew()`, :meth:`Rolling.kurt()`, :meth:`Expanding.skew()` and :meth:`Expanding.kurt()` through implementation of Kahan summation (:issue:`6929`)
 - Improved error reporting for subsetting columns of a :class:`DataFrameGroupBy` with ``axis=1`` (:issue:`37725`)
+- Add support for ``min_count`` keyword for :meth:`DataFrame.groupby` and :meth:`DataFrame.resample` for functions ``min``, ``max``, ``first`` and ``last`` (:issue:`37821`, :issue:`37768`)
```
Comment (rhshadrach) [inline doc/source/whatsnew/v1.2.0.rst:238]: And nth, right?

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Overall looks really good, just some minor comments on tests.

File: pandas/_libs/groupby.pyx (hunk 2 above)
Comment (phofl) [reply pandas/_libs/groupby.pyx:1059]: Indirectly through cython tests, but is better to add this too. Adjusted the test to reflect the behavior

File: doc/source/whatsnew/v1.2.0.rst (hunk 3 above)
Comment (phofl) [reply doc/source/whatsnew/v1.2.0.rst:238]: No, the cython function yes, but the function does not accept min_count. Have only added suppport for functions which have min count documented in this first step

File: pandas/tests/groupby/aggregate/test_other.py (hunk 1 above)
Comment (phofl) [reply pandas/tests/groupby/aggregate/test_other.py]: Thx, moved and renamed it

**Review by rhshadrach — APPROVED** [review]
cc @jreback @jbrockmendel

File: pandas/_libs/groupby.pyx [hunk 4]
```diff
@@ -1411,7 +1407,7 @@ def group_min(groupby_t[:, :] out,
 
         for i in range(ncounts):
             for j in range(K):
-                if nobs[i, j] == 0:
+                if nobs[i, j] < min_count or nobs[i, j] == 0:
```
Comment (jbrockmendel) [inline pandas/_libs/groupby.pyx:1410]: can you avoid the double lookup/comparison here by setting `min_count = max(min_count, 1)` outside the loop?

Comment (phofl) [reply pandas/_libs/groupby.pyx]: Done

**Comment by jreback** [comment]
lgtm. ping on green (resolved conflict)

**Comment by phofl** [comment]
@jreback greenish. Failure is unrelated

**Comment by jreback** [comment]
thanks @phofl

**Pull request merged**